File: prime-api/app/baldomar_validation.py
"""
Validation rules aligned with Baldomar,Galapate_Final Project (signup_process.php + validation.js).
"""

from __future__ import annotations

import re
from typing import Optional

MAX_PASSWORD_UTF8_BYTES = 72
MIN_PASSWORD_UTF8_BYTES = 8
_PASSWORD_SPECIAL_RE = re.compile(r'[!@#$%^&*(),.?":{}|<>]')


def validate_person_name(name: str, field: str = "Name") -> Optional[str]:
    """Letters, spaces, hyphen, apostrophe; title case per word; no ALL CAPS / double spaces / triple letters."""
    name = (name or "").strip()
    if not name:
        return f"{field} is required."
    if re.search(r"[^a-zA-Z\s\-']", name):
        return f"{field} contains invalid characters."
    if re.search(r"[A-Z]{2,}", name) and name.upper() == name and re.search(r"[A-Z]", name):
        return f"{field} must not be ALL CAPS."
    if "  " in name:
        return f"{field} must not contain double spaces."
    if re.search(r"(.)\1{2,}", name, re.IGNORECASE):
        return f"{field} must not contain three (3) consecutive identical letters."
    for w in name.split():
        if not w:
            continue
        first, rest = w[0], w[1:]
        if first.upper() != first or (rest and rest.lower() != rest):
            return f"{field} must be Capitalized properly. Example: Juan Carlo"
    return None


def validate_address_line(value: str, field: str) -> Optional[str]:
    """Street-style lines: alnum, spaces, hyphen, comma, period, apostrophe; capital start; no double spaces."""
    v = (value or "").strip()
    if not v:
        return f"{field} is required."
    if re.search(r"[^A-Za-z0-9\s\-\.,']", v):
        return f"{field} contains invalid characters."
    if "  " in v:
        return f"{field} must not contain double spaces."
    if re.search(r"(.)\1{2,}", v, re.IGNORECASE):
        return f"{field} must not contain three (3) consecutive identical letters."
    first = v[0]
    if first != first.upper() or not re.search(r"[A-Z]", first):
        return f"{field} must start with a capital letter."
    return None


def validate_password_baldomar(password: str) -> Optional[str]:
    """Min 8 UTF-8 bytes, max 72 (bcrypt); must include at least one number and one special character."""
    raw = password.encode("utf-8")
    if len(raw) < MIN_PASSWORD_UTF8_BYTES:
        return "Password must be at least 8 characters."
    if len(raw) > MAX_PASSWORD_UTF8_BYTES:
        return "Password must be at most 72 bytes (bcrypt limit)."
    if not re.search(r"[0-9]", password):
        return "Password must include at least one number (0-9)."
    if not _PASSWORD_SPECIAL_RE.search(password):
        return (
            "Password must include at least one special character "
            '(e.g. ! @ # $ % ^ & * ( ) , . ? " : { } | < >).'
        )
    return None


def validate_optional_security_answer(value: Optional[str], field: str) -> Optional[str]:
    if value is None or value == "":
        return None
    s = value.strip()
    if not s:
        return f"{field} cannot be only whitespace."
    if len(s) > 2000:
        return f"{field} is too long."
    return None


def validate_gender_optional(value: Optional[str]) -> Optional[str]:
    if value is None or value == "":
        return None
    if value not in ("Male", "Female"):
        return "Gender must be Male or Female."
    return None

File: prime-next/src/lib/baldomarValidation.ts
/**
 * Client-side mirrors of Baldomar,Galapate_Final Project + prime-api baldomar_validation.py
 */

export const MIN_PASSWORD_UTF8_BYTES = 8;
export const MAX_PASSWORD_UTF8_BYTES = 72;

export function validatePersonName(name: string, field = "Name"): string | null {
  const n = name.trim();
  if (!n) return `${field} is required.`;
  if (/[^a-zA-Z\s\-']/.test(n)) return `${field} contains invalid characters.`;
  if (/[A-Z]{2,}/.test(n) && n.toUpperCase() === n && /[A-Z]/.test(n)) {
    return `${field} must not be ALL CAPS.`;
  }
  if (/\s{2,}/.test(n)) return `${field} must not contain double spaces.`;
  if (/(.)\1{2,}/i.test(n)) {
    return `${field} must not contain three (3) consecutive identical letters.`;
  }
  const words = n.split(/\s+/);
  for (const w of words) {
    if (!w) continue;
    const first = w.charAt(0);
    const rest = w.slice(1);
    if (first !== first.toUpperCase() || (rest && rest !== rest.toLowerCase())) {
      return `${field} must be Capitalized properly. Example: Juan Carlo`;
    }
  }
  return null;
}

export function validateAddressLine(value: string, field: string): string | null {
  const v = value.trim();
  if (!v) return `${field} is required.`;
  if (/[^A-Za-z0-9\s\-\.,']/.test(v)) return `${field} contains invalid characters.`;
  if (/\s{2,}/.test(v)) return `${field} must not contain double spaces.`;
  if (/(.)\1{2,}/i.test(v)) {
    return `${field} must not contain three (3) consecutive identical letters.`;
  }
  const first = v.charAt(0);
  if (first !== first.toUpperCase() || !/[A-Z]/.test(first)) {
    return `${field} must start with a capital letter.`;
  }
  return null;
}

const PASSWORD_SPECIAL_RE = /[!@#$%^&*(),.?":{}|<>]/;

export function validatePasswordBaldomar(password: string): string | null {
  const enc = new TextEncoder().encode(password);
  if (enc.length < MIN_PASSWORD_UTF8_BYTES) {
    return "Password must be at least 8 characters.";
  }
  if (enc.length > MAX_PASSWORD_UTF8_BYTES) {
    return "Password must be at most 72 bytes (bcrypt limit).";
  }
  if (!/[0-9]/.test(password)) {
    return "Password must include at least one number (0-9).";
  }
  if (!PASSWORD_SPECIAL_RE.test(password)) {
    return (
      'Password must include at least one special character ' +
      '(e.g. ! @ # $ % ^ & * ( ) , . ? " : { } | < >).'
    );
  }
  return null;
}

export function validateOptionalSecurityAnswer(
  value: string | null | undefined,
  field: string,
): string | null {
  if (value == null || value === "") return null;
  const s = value.trim();
  if (!s) return `${field} cannot be only whitespace.`;
  if (s.length > 2000) return `${field} is too long.`;
  return null;
}

export function validateGender(value: string | null | undefined): string | null {
  if (value == null || value === "") return null;
  if (value !== "Male" && value !== "Female") {
    return "Gender must be Male or Female.";
  }
  return null;
}

export function validateRegisterForm(input: {
  fullName: string;
  email: string;
  password: string;
  gender: string;
  securityQ1: string;
  securityQ2: string;
}): string | null {
  const nameErr = validatePersonName(input.fullName, "Full name");
  if (nameErr) return nameErr;
  if (!input.gender?.trim()) return "Gender is required.";
  const genderErr = validateGender(input.gender);
  if (genderErr) return genderErr;
  const pwErr = validatePasswordBaldomar(input.password);
  if (pwErr) return pwErr;
  if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(input.email.trim())) {
    return "Please enter a valid email address.";
  }
  if (!input.securityQ1?.trim()) return "Security answer 1 is required.";
  if (!input.securityQ2?.trim()) return "Security answer 2 is required.";
  const s1 = validateOptionalSecurityAnswer(input.securityQ1, "Security answer 1");
  if (s1) return s1;
  const s2 = validateOptionalSecurityAnswer(input.securityQ2, "Security answer 2");
  if (s2) return s2;
  return null;
}

export function validateAppointmentBaldomar(input: {
  fullName: string;
  address: string;
  location: string;
  age: number;
}): string | null {
  const nameErr = validatePersonName(input.fullName, "Name");
  if (nameErr) return nameErr;
  if (!Number.isFinite(input.age) || input.age < 18) {
    return "You must be 18 years or older.";
  }
  const addrErr = validateAddressLine(input.address, "Address");
  if (addrErr) return addrErr;
  const locErr = validateAddressLine(input.location, "Location");
  if (locErr) return locErr;
  return null;
}
